### ISSUE TITLE
Callbacks are not compatible with Xamarin.iOS

### ISSUE THREAD

**Issue by Clancey** [ManagedBass/ManagedBass#44]
System.ExecutionEngineException: Attempting to JIT compile method '(wrapper native-to-managed) MusicPlayer.BassPlayer:OnFileClose (intptr)' while running in aot-only mode. See https://developer.xamarin.com/guides/ios/advanced_topics/limitations/ for more information.

  at (wrapper managed-to-native) System.Object:__icall_wrapper_mono_delegate_to_ftnptr (object)
  at (wrapper managed-to-native) ManagedBass.Bass:BASS_StreamCreateFileUser (ManagedBass.StreamSystem,ManagedBass.BassFlags,ManagedBass.FileProcedures,intptr)
  at ManagedBass.Bass.CreateStream (ManagedBass.StreamSystem System, ManagedBass.BassFlags Flags, ManagedBass.FileProcedures Procedures, System.IntPtr User) [0x00000] in <71311c6a11824b2fa322046a464e4feb>:0

**Comment by MathewSachin** [comment]
Please make sure the callbacks you are passing as instances of `FileProcedures` are originally defined as `static` methods with `MonoPInvokeCallback` attribute applied.

**Comment by Clancey** [comment]
Correct
```
[MonoPInvokeCallback(typeof(FileReadProcedure))]
static int OnFileRead(IntPtr buffer, int length, IntPtr user)
```

Same issue on my End of track sync proc: 

```
[MonoPInvokeCallback(typeof(SyncProcedure))]
public static void OnSyncProc(int handle, int channel, int data, IntPtr user)
```


```
System.ExecutionEngineException: Attempting to JIT compile method '(wrapper native-to-managed) ManagedBass.ChannelReferences:Callback (int,int,int,intptr)' while running in aot-only mode. See https://developer.xamarin.com/guides/ios/advanced_topics/limitations/ for more information.

  at (wrapper managed-to-native) System.Object:__icall_wrapper_mono_delegate_to_ftnptr (object)
  at (wrapper managed-to-native) ManagedBass.Bass:BASS_ChannelSetSync (int,ManagedBass.SyncFlags,long,ManagedBass.SyncProcedure,intptr)
  at ManagedBass.Bass.ChannelSetSync (System.Int32 Handle, ManagedBass.SyncFlags Type, System.Int64 Parameter, ManagedBass.SyncProcedure Procedure, System.IntPtr User) [0x00039] in <71311c6a11824b2fa322046a464e4feb>:0 
  at ManagedBass.ChannelReferences.Add (System.Int32 Handle, System.Int32 SpecificHandle, System.Object proc) [0x00054] in <71311c6a11824b2fa322046a464e4feb>:0 
  at ManagedBass.Bass.ChannelSetSync (System.Int32 Handle, ManagedBass.SyncFlags Type, System.Int64 Parameter, ManagedBass.SyncProcedure Procedure, System.IntPtr User) [0x00048] in <71311c6a11824b2fa322046a464e4feb>:0 
```

**Comment by MathewSachin** [comment]
I think the issue might be happening due to the callback present in ManagedBass itself.

https://github.com/ManagedBass/ManagedBass/blob/bcd197f35d7d5a495a226be5a93643674df80b92/src/Bass/Shared/ChannelReferences.cs#L47

I'll try removing it when compiling for iOS.

**Comment by Clancey** [comment]
Yes, that is causing it for the SyncProc. Is there a reason you are double hooking it up and keeping a second managed reference? That just sounds expensive.

**Comment by Clancey** [comment]
Also looking at the full stack for the FileProc, it is the same thing.

**Comment by MathewSachin** [comment]
We are keeping a second managed reference because BASS fails with error if the delegate gets garbage collected.

But, I think this is not required for iOS.

**Comment by Clancey** [comment]
Sure, but that is normal with anything in .net. I am just used to keeping my references for things I need. :D

**Comment by MathewSachin** [comment]
Please test with the attached DLL:

[ManagedBass.zip](https://github.com/ManagedBass/ManagedBass/files/1436629/ManagedBass.zip)

**Comment by Clancey** [comment]
That works perfectly!  Any chance of a nuget update? Also the BassFX Nuget as well? https://github.com/ManagedBass/ManagedBass/issues/43

**Comment by MathewSachin** [comment]
Will do soon

**Comment by Clancey** [comment]
Fixed in the latest 2.0.4 nuget. Closing the issue.
